### PR TITLE
perf(dflash2): fuse two-tap grouped convolution

### PR DESCRIPTION
**Pull request by voipmonitor**
## Resulting behavior

GLM-5.3 DFlash2 executes each supported two-tap grouped convolution with one
Triton kernel instead of materializing coefficient, shift, mask, multiply, and
reduction intermediates as separate PyTorch launches.

The fused path requires CUDA BF16 tensors, contiguous hidden and base tensors,
two taps, groups of 16 channels, an eight-token recurrent block, and an exact
`(rows, taps, groups)` delta tensor. The GLM-5.3 checkpoint geometry is 4,096
hidden channels and 256 groups. Broadcastable row, tap, or group delta tensors
retain the generic implementation.

Explicit BF16 round-to-nearest boundaries preserve the generic operation order.
The optimized output is bit-identical to the BF16 reference, including recurrent
block boundaries where the predecessor tap must be zero.

The implementation commit retains MadeBy561 as its author. The tests and fused
dispatch safety guard are authored by Martin Vit.

## Compatibility

Checkpoint formats and serving arguments are unchanged. Unsupported dtypes,
layouts, shapes, and broadcastable delta inputs use the existing generic path.

## Validation

The CUDA suite ran in the source-locked GLM-5.3 runtime on one NVIDIA RTX PRO
6000 Blackwell Workstation GPU:

```text
CUDA_VISIBLE_DEVICES=0 /opt/venv/bin/python -m pytest \
  --confcutdir=/tmp /tmp/test_dflash2_grouped_conv.py -vv
11 passed
```

Exact equality covers 1, 7, 8, 9, 16, 17, and 64 rows; row, tap, and group
broadcasting; and CUDA graph capture/replay. A graph-amortized launch sweep over
blocks of 128, 256, 512, and 1,024 elements selected 256 for the production
8-, 32-, and 64-row shapes.

An isolated TP4/DCP1 comparison used the same GLM-5.3 NVFP4 target, seven-token
MXFP8 DFlash2 draft, independent 512-token target and recurrent cache pages, a
4,096-token scheduler budget, full decode CUDA graphs, and four RTX PRO 6000
Blackwell Workstation GPUs with a +6000 MHz memory-clock offset:

| Concurrency | Generic grouped convolution | Fused grouped convolution | Change |
| --- | ---: | ---: | ---: |
| 1 | 87.49 verifier steps/s | 90.07 verifier steps/s | +2.95% |
| 8 | 304.64 verifier steps/s | 306.54 verifier steps/s | +0.62% |

Searches of open `vllm-project/vllm` and `local-inference-lab/vllm` pull
requests found no other DFlash2 grouped-convolution implementation.

AI assistance was used to inspect the implementation, construct tests, analyze
review findings, and prepare this pull request. The human submitter reviewed the
changed lines and validation evidence.

### CODE REVIEW

**Comment by voipmonitor** [comment]
@coderabbitai review

File: tests/kernels/test_dflash2_grouped_conv.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+import pytest
+import torch
+import torch.nn.functional as F
+
+from vllm.model_executor.models.qwen3_dflash2 import _grouped_conv
+
+pytestmark = pytest.mark.skipif(
+    not torch.accelerator.is_available(),
+    reason="The fused grouped convolution requires CUDA",
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply tests/kernels/test_dflash2_grouped_conv.py:13]: Resolved. The module now skips on `torch.cuda.is_available()`. The CUDA-only suite passes 11/11 on an RTX PRO 6000 Blackwell.

File: vllm/model_executor/models/qwen3_dflash2.py [hunk 2]
```diff
@@ -29,6 +144,30 @@ def _grouped_conv(
...
+    if (
+        hidden_states.is_cuda
+        and hidden_states.dtype == torch.bfloat16
+        and delta.dtype == torch.bfloat16
+        and base.dtype == torch.bfloat16
+        and hidden_states.ndim == 2
+        and hidden_states.is_contiguous()
+        and base.is_contiguous()
+        and delta.ndim == 3
+        and delta.stride(2) == 1
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply vllm/model_executor/models/qwen3_dflash2.py:157]: Resolved. Fused dispatch now requires the exact `(rows, taps, groups)` delta shape. Row-, tap-, and group-broadcast inputs are explicitly tested through the generic path; all 11 CUDA tests pass.

**Comment by voipmonitor** [comment]
R26 packaged-integration validation: this pull request is included in `voipmonitor/vllm:jovian-judgement-community-20260905-r26` (`sha256:d0592ea9d73cac5aadb151a58bbb43cf7aff03829d46bb4f4ba7396aaef67c68`). The exact two-layer image passed 383 focused tests with 33 unsupported-device skips. On four stock RTX PRO 6000 Blackwell Workstation Edition GPUs, the packaged runtime measured 14,937 tok/s and 170.1 tok/s C1 for no speculation, 14,509 tok/s and 109.0 verifier steps/s for MTP3, and 12,783 tok/s plus 81.3 verifier steps/s for DFlash2 DCP4 full-CKV. This is full-stack compatibility and regression evidence; isolated performance claims remain those documented in this pull request. Merge order and exact source mirrors are recorded in #651.

**Comment by voipmonitor** [comment]
## R27 integration validation

The change represented by this PR is included in the qualified, source-locked
GLM-5.3-Flash runtime
`voipmonitor/vllm:jovian-judgement-community-20260906-r27`
(`sha256:a298fe1cd207eaf97bd2ff2686716ed25b7009c09b36650eba732a4a7dc51512`).
The exact vLLM composition is mirrored at
[`voipmonitor/vllm:integration/glm53-r27-release-20260906`](https://github.com/voipmonitor/vllm/tree/integration/glm53-r27-release-20260906),
commit `63a82f8d323e8538cbe6f88ae1812a1c01577a0f`.

Qualification used four stock-clock RTX PRO 6000 Blackwell Workstation Edition
GPUs, TP4, a 4,096-token scheduler budget, 16 NCCL channels, a 2 MiB NCCL
buffer, and full plus piecewise CUDA graphs:

| Mode | DCP | 32K prefill | C1 output / steps | C8 output / steps |
|---|---:|---:|---:|---:|
| No speculation | 1 | 14,870 tok/s | 170.6 tok/s | 733.8 tok/s |
| MTP3 | 1 | 14,468 tok/s | 276.0 / 109.1 tok/s | 901.0 / 371.3 tok/s |
| MTP3 full CKV | 4 | 12,864 tok/s | 247.0 / 97.1 tok/s | 876.4 / 346.4 tok/s |
| DFlash2 K7 full CKV, NVFP4 KV | 4 | 12,633 tok/s | 198.0 / 81.2 tok/s | 645.5 / 260.8 tok/s |

FP8 no-speculation and NVFP4 DFlash2 external-cache configurations also passed
cold compute, vLLM prefix reuse, engine-driven RAM-L1 restore, full-process
filesystem-L2 restore, and block-checksum validation on all four ranks. An exact
81,576-token leading-instruction test reused 81,567 tokens when only the user
continuation changed.

This is an integration and regression gate, not an isolated attribution of the
aggregate throughput to this PR. The complete open-PR merge order and evidence
are recorded in #651.

**Comment by voipmonitor** [comment]
Included in `dev/jovian-judgement` through [this PR's individual merge](https://github.com/local-inference-lab/vllm/commit/0f77d04485cbb67094a2d6b34055b68ad123c277). The reviewed head and its contributor commits remain ancestors; the PR is merged and closed.

Source validation: replaying all 32 R35 review heads on the pinned base exactly reproduces the released Docker's vLLM tree; all 6,870 installed tracked files match. JJ additionally preserves Luke's DS4.1 work and #734. The final composition passed 247 focused checkpoint/scheduler, sampler/warmup and native GPU tests. This is combined-source evidence, not a fresh performance or full-model qualification for this individual PR.

Publication-history clarification: the individual merge linked above is in JJ's first-parent history. It replaces the receipt's archived wrapper-merge reference; GitHub's historical merge SHA may still identify that archive. See #731 for component review order and qualification limits.
